design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,287 @@
+running a Kubernetes cluster. This document proposes a common pattern for stacks
+that utilize the Kubernetes client as their API. Motivation for this design
+comes from the desire to support [Rook] as a provider in Crossplane.
+
+Using the Kubernetes API to create objects in a controller's reconcile loop is
+not a foreign concept in Crossplane. Existing claim reconcilers create a
+Kubernetes object to represent the external resource. However, Kubernetes-native
+providers differ in that the external resource is actually a Kubernetes object
+itself. This concept is also not new in Crossplane, as [complex workloads]
+operate in this manner. `KubernetesApplication` instances are first scheduled to
+a Kubernetes cluster by the scheduler controller, which causes them to pass the
+predicates of the application controller. It uses the template(s) in the
+`KubernetesApplication` instance to create one or more
+`KubernetesApplicationResource` objects, which are then picked up by the
+resource scheduler. It is responsible for actually creating the
+Kubernetes-native (e.g. `Deployment`, `ConfigMap`, etc.) objects in the target
+cluster, then managing them as a managed reconciler does for an external
+resource. In this scenario, the Kubernetes-native objects are the external
+resources.
+
+## Goals
+
+* Create a common pattern that can be used across infrastructure stacks (and
+  potentially the existing complex workload controllers) that utilize the
+  Kubernetes API as the infrastructure provider API
+* Demonstrate the usage of this pattern with Rook
+
+## Differences from `KubernetesApplication`
+
+While the `KubernetesApplication` pattern detailed above provides a starting
+point for this design, a key difference is that `KubernetesApplication` does not
+adhere to the separation of concern structure of other Crossplane resources
+because Kubernetes resources are not modeled in a `claim <-> portable class <->
+non-portable class <-> managed <-> external` workflow. Kubernetes-native
+infrastructure resources, on the other hand, *should* be modeled with the same
+separation of concern that other cloud-provider resources expose.
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to provision [Minio] object storage in a
+Kubernetes cluster. Assuming you have installed Rook and the corresponding Minio
+operator into a cluster, you should be able to dynamically provision a Minio
+object storage cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a Kubernetes `Provider` for target cluster:
```

Comment:
i wonder if there may be a nice crossplane CLI command that could make this easy, using the locally available kubeconfig contexts.  e.g.:

```
kubectl crossplane create provider --namespace=rook-infra-dev --kubectx=on-prem-1
```